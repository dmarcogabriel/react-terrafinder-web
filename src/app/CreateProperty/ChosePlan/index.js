import { ChosePlan } from './ChosePlan';

export { ChosePlan };
