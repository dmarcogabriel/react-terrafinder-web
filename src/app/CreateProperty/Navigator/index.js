import React from 'react';
import Button from 'common/components/Button';
import classes from './Navigator.module.scss';

export default function Navigator({
  onBack,
  onNext,
  nextButtonText = 'Pr√≥ximo',
}) {
  return (
    <div className={classes.navigator}>
      <button
        data-testid="back"
        className={classes.backButton}
        type="button"
        onClick={onBack}
      >
        Voltar
      </button>

      <Button dataTestId="next" onClick={onNext} className={classes.nextButton}>
        {nextButtonText}
      </Button>
    </div>
  );
}
