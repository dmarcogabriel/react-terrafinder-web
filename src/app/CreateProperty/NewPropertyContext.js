import React, { createContext, useContext, useState } from 'react';

const Context = createContext({});

export const useNewProperty = () => useContext(Context);

export const NewPropertyProvider = ({ children }) => {
  const [property, setProperty] = useState({});
  const [step, setStep] = useState(1);
  const [photos, setPhotos] = useState([]);

  const updateProperty = (data) => setProperty({ ...property, ...data });

  const nextStep = () => setStep(step + 1);

  React.useEffect(() => {
    console.log('STEP', step);
  }, [step]);

  React.useEffect(() => {
    console.log('PROPERTY', property);
  }, [property]);

  return (
    <Context.Provider
      value={{
        property,
        updateProperty,
        nextStep,
        step,
        photos,
        setPhotos,
      }}
    >
      {children}
    </Context.Provider>
  );
};
