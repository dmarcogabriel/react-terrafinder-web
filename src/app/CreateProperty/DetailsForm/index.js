import React, { useState } from 'react';
import Input from 'common/components/Input';
import { useLocation, useHistory } from 'react-router-dom';
import api from 'services/api';
import { useFormik } from 'formik';
import { object, string } from 'yup';
import { MdAdd } from 'react-icons/md';
import Button from 'common/components/atm/Button';
import { validateArrayOfInputs } from 'utils/validators';
import { useUser } from 'hooks/useUser';
import { useNotification, NOTIFICATION_TYPES } from 'hooks/useNotification';
import classes from './DetailsForm.module.scss';
import Navigator from '../Navigator';

export default function DetailsForm() {
  const { state } = useLocation();
  const [farming, setFarming] = useState([{ value: '', error: '' }]);
  const [activities, setActivities] = useState([{ value: '', error: '' }]);
  const history = useHistory();
  const { currentUser } = useUser();
  const { showNotification } = useNotification();

  const { values, errors, handleChange, handleSubmit } = useFormik({
    initialValues: {
      amount: '',
      size: '',
    },
    validationSchema: object().shape({
      amount: string().required('Campo obrigatório'),
      size: string().required('Campo obrigatório'),
    }),
    async onSubmit() {
      if (validateArrayOfInputs(activities) && validateArrayOfInputs(farming)) {
        try {
          const res = await api.post(
            'properties',
            {
              ...values,
              ...state.values,
              user: currentUser._id,
            },
            {
              headers: {
                'x-access-token': currentUser.token,
              },
            }
          );

          showNotification(
            'Propriedade criada com sucesso!',
            NOTIFICATION_TYPES.SUCCESS
          );

          history.push(
            `/create/property/upload-photos/${res.data.property._id}?step=3`
          );
        } catch (err) {
          showNotification(
            'Ops, ocorreu um erro na operação',
            NOTIFICATION_TYPES.ERROR
          );
        }
      } else {
        showNotification(
          'Ops, ocorreu um erro na operação',
          NOTIFICATION_TYPES.ERROR
        );
      }
    },
  });

  const addActivity = () => {
    if (activities.length < 5)
      setActivities((old) => [...old, { value: '', error: '' }]);
  };

  const addFarming = () => {
    if (farming.length < 5)
      setFarming((old) => [...old, { value: '', error: '' }]);
  };

  const handleChangeActivity = (e, i) => {
    setActivities(
      activities.map((act, index) => (i === index ? { value: e } : act))
    );
  };

  const handleChangeFarm = (e, i) => {
    setFarming(
      farming.map((farm, index) => (i === index ? { value: e } : farm))
    );
  };

  const goBack = () => history.goBack();

  return (
    <div className={classes.detailsForm}>
      <div className={classes.inlineInputs}>
        <Input
          label="Quanto quer pela propriedade?"
          type="number"
          value={values.amount}
          onChange={handleChange('amount')}
          errorMessage={errors.amount}
        />
        <Input
          label="Qual o tamanho da propriedade (em equitares)"
          value={values.size}
          onChange={handleChange('size')}
          errorMessage={errors.size}
        />
      </div>

      <div className={classes.inlineInputs}>
        <div className={classes.col}>
          <p>Cultivo (plantações)</p>

          <Button
            modifiers="success"
            dataTestId="addFarming"
            onClick={addFarming}
          >
            <MdAdd size={22} />
          </Button>

          <div data-testid="farms">
            {farming.map((farm, i) => (
              <div data-testid={`farm-${i}`} key={String(i)}>
                <Input
                  dataTestId={`farmInput-${i}`}
                  key={String(i)}
                  value={farm.value}
                  onChange={(e) => handleChangeFarm(e, i)}
                  errorMessage={farm.error}
                />
              </div>
            ))}
          </div>
        </div>

        <div className={classes.col}>
          <p>Atividades e Destaques</p>

          <Button
            modifiers="success"
            dataTestId="actButton"
            onClick={addActivity}
          >
            <MdAdd size={22} />
          </Button>

          <div data-testid="acts">
            {activities.map((act, i) => (
              <div key={String(i)} data-testid={`act-${i}`}>
                <Input
                  dataTestId={`actInput-${i}`}
                  key={String(i)}
                  value={act.value}
                  onChange={(e) => handleChangeActivity(e, i)}
                  errorMessage={act.error}
                />
              </div>
            ))}
          </div>
        </div>
      </div>

      <Navigator
        onBack={goBack}
        onNext={handleSubmit}
        nextButtonText="Cadastrar"
      />
    </div>
  );
}
