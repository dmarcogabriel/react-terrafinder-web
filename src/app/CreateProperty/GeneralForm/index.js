import React from 'react';
import Input from 'common/components/Input';
import TextArea from 'common/components/TextArea';
import Select from 'common/components/Select';
import { useFormik } from 'formik';
import { object, string } from 'yup';
import { useUser } from 'hooks/useUser';
import { useHistory } from 'react-router-dom';
import classes from './GeneralForm.module.scss';
import Navigator from '../Navigator';
import PROPERTY_KINDS from './propertyKinds';

export default function GeneralForm() {
  const { currentUser } = useUser();
  const history = useHistory();

  const { values, handleChange, handleSubmit, errors } = useFormik({
    initialValues: {
      name: '',
      ownerName: currentUser.firstName || '',
      description: '',
      propertyKind: '',
      state: '',
      nearbyCity: '',
      cep: '',
    },
    validationSchema: object().shape({
      name: string().required('Campo obrigatório'),
      ownerName: string().required('Campo obrigatório'),
      description: string().required('Campo obrigatório'),
      propertyKind: string().required('Campo obrigatório'),
      state: string().required('Campo obrigatório'),
      nearbyCity: string().required('Campo obrigatório'),
      cep: string().required('Campo obrigatório'),
    }),
    async onSubmit() {
      history.push('/create/property/details?step=2', { values });
    },
  });

  return (
    <div className={classes.generalForm}>
      <div className={classes.inlineInputs}>
        <Input
          label="Nome da Propriedade"
          value={values.name}
          onChange={handleChange('name')}
          errorMessage={errors.name}
        />
        <Input
          label="Nome do Proprietário"
          value={values.ownerName}
          onChange={handleChange('ownerName')}
          errorMessage={errors.ownerName}
        />
      </div>

      <TextArea
        label="Descrição"
        value={values.description}
        onChange={handleChange('description')}
        errorMessage={errors.description}
      />
      <Select
        label="Tipo de Imóvel"
        options={PROPERTY_KINDS}
        value={values.propertyKind}
        onChange={(e) => handleChange('propertyKind')(e.name)}
        errorMessage={errors.propertyKind}
      />

      <div className={classes.inlineInputs}>
        <Input
          label="Estado (em que o imóvel se encontra)"
          value={values.state}
          onChange={handleChange('state')}
          maxLength={2}
          errorMessage={errors.state}
        />
        <Input
          label="Cidade Mais Próxima ao imóvel"
          value={values.nearbyCity}
          onChange={handleChange('nearbyCity')}
          errorMessage={errors.nearbyCity}
        />
      </div>

      <Input
        label="CEP da Propriedade"
        value={values.cep}
        onChange={handleChange('cep')}
        errorMessage={errors.cep}
      />

      <Navigator onBack={history.goBack} onNext={handleSubmit} />
    </div>
  );
}
