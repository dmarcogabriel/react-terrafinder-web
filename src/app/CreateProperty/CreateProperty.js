import React from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  useRouteMatch,
  useHistory,
} from 'react-router-dom';
import {
  CreatePropertyProgress,
  ChosePlan,
  BasicInfoForm,
  DetailsForm,
  UploadPhotos,
  Review,
} from './components';
import { NewPropertyProvider } from './NewPropertyContext';

export const CreateProperty = () => {
  const { path } = useRouteMatch();
  const history = useHistory();

  return (
    <NewPropertyProvider>
      <CreatePropertyProgress currentStep={1}>
        <Router>
          <Switch>
            <Route exact path={path}>
              <ChosePlan />
            </Route>
            <Route path={`${path}/basic-info`}>
              <BasicInfoForm />
            </Route>
            <Route path={`${path}/details`}>
              <DetailsForm />
            </Route>
            <Route path={`${path}/upload-photos`}>
              <UploadPhotos />
            </Route>
            <Route path={`${path}/review`}>
              <Review history={history} />
            </Route>
          </Switch>
        </Router>
      </CreatePropertyProgress>
    </NewPropertyProvider>
  );
};
