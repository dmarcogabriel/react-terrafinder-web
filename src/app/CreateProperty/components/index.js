import CreatePropertyProgress from './CreatePropertyProgress';
import ChosePlan from './Steps/ChosePlan';
import BasicInfoForm from './Steps/BasicInfoForm';
import DetailsForm from './Steps/DetailsForm';
import UploadPhotos from './Steps/UploadPhotos';
import Review from './Steps/Review';

export {
  CreatePropertyProgress,
  ChosePlan,
  BasicInfoForm,
  DetailsForm,
  UploadPhotos,
  Review,
};
