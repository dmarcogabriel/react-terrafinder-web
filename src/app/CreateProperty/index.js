import { GeneralForm } from './GeneralForm';
import { DetailsForm } from './DetailsForm';
import { UploadPhotos } from './UploadPhotos';

export { GeneralForm, DetailsForm, UploadPhotos };
