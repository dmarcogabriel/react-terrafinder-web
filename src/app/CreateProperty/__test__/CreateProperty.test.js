import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
// import { render } from '@testing-library/react';
import { UserProvider } from 'contexts/User';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import CreateProperty from '..';

const history = createMemoryHistory();

const Comp = (props) => (
  <UserProvider>
    <Router history={history}>
      <CreateProperty {...props} />
    </Router>
  </UserProvider>
);

describe('<CreateProperty />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Comp />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  // Add tests here...

  it('matches snapshot', () => {
    const tree = renderer.create(<Comp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
