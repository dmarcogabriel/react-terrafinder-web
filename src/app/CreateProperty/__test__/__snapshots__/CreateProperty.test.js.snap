// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProperty /> matches snapshot 1`] = `
<div>
  <div>
    <h1>
      Anúncio de Imóvel Rural
    </h1>
  </div>
  <div>
    <div>
      <div>
        <p>
          1
        </p>
        <div>
          <p>
            Informações Gerais
          </p>
          <p>
            Nos fale um pouco mais sobre o seu imóvel
          </p>
        </div>
      </div>
      <div>
        <p>
          2
        </p>
        <div>
          <p>
            Informações Detalhadas
          </p>
          <p>
            Conte-nos mais detalhes sobre sua propriedade, para chamar mais a atenção
          </p>
        </div>
      </div>
      <div>
        <p>
          3
        </p>
        <div>
          <p>
            Conteúdos
          </p>
          <p>
            Você pode enviar imagens e vídeos da sua propriedade
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
